pulp_container/app/migrations/0030_enforce_tagged_manifest_reference.py

```diff
@@ -0,0 +1,25 @@
+# Generated by Django 3.2.13 on 2022-05-16 16:13
+
+from django.db import migrations, models
+import django.db.models.deletion
+
+
+def remove_tags_without_manifests(apps, schema_editor):
+    Tag = apps.get_model("container", "Tag")
+    Tag.objects.filter(tagged_manifest=None).delete()
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('container', '0029_remove_blob_media_type'),
+    ]
+
+    operations = [
+        migrations.RunPython(code=remove_tags_without_manifests),
+        migrations.AlterField(
```

Comment:
Yes, it works. In the beginning, I had the `RunPython` call after `AlterField`, that is why the migration failed:
```
IPython 8.2.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: Tag.objects.create(name="test")
Out[1]: <Tag: test>

In [2]: Tag.objects.all()
Out[2]: <BulkTouchQuerySet [<Tag: latest>, <Tag: test>]>

In [3]:                                                                                                               
Do you really want to exit ([y]/n)? 
(pulp) [vagrant@pulp3-source-fedora35 ~]$ django-admin makemigrations --name enforce_tagged_manifest_reference
You are trying to change the nullable field 'tagged_manifest' on tag to non-nullable without a default; we can't do that (the database needs something to populate existing rows).
Please select a fix:
 1) Provide a one-off default now (will be set on all existing rows with a null value for this column)
 2) Ignore for now, and let me handle existing rows with NULL myself (e.g. because you added a RunPython or RunSQL operation to handle NULL values in a previous data migration)
 3) Quit, and let me add a default in models.py
Select an option: 2
Migrations for 'container':
  devel/pulp_container/pulp_container/app/migrations/0030_enforce_tagged_manifest_reference.py
    - Alter field tagged_manifest on tag
(pulp) [vagrant@pulp3-source-fedora35 ~]$ vim devel/pulp_container/pulp_container/app/migrations/0030_enforce_tagged_manifest_reference.py 
(pulp) [vagrant@pulp3-source-fedora35 ~]$ django-admin migrate
Operations to perform:
  Apply all migrations: auth, container, contenttypes, core, file, ostree, sessions
Running migrations:
  Applying container.0030_enforce_tagged_manifest_reference...Traceback (most recent call last):
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.NotNullViolation: column "tagged_manifest_id" of relation "container_tag" contains null values


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/pulp/bin/django-admin", line 8, in <module>
    sys.exit(execute_from_command_line())
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/core/management/base.py", line 89, in wrapped
    res = handle_func(*args, **kwargs)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/core/management/commands/migrate.py", line 244, in handle
    post_migrate_state = executor.migrate(
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/db/migrations/executor.py", line 117, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/db/migrations/executor.py", line 147, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/db/migrations/executor.py", line 227, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/db/migrations/migration.py", line 126, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/db/migrations/operations/fields.py", line 244, in database_forwards
    schema_editor.alter_field(from_model, from_field, to_field)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/db/backends/base/schema.py", line 608, in alter_field
    self._alter_field(model, old_field, new_field, old_type, new_type,
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/db/backends/postgresql/schema.py", line 196, in _alter_field
    super()._alter_field(
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/db/backends/base/schema.py", line 765, in _alter_field
    self.execute(
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/db/backends/base/schema.py", line 145, in execute
    cursor.execute(sql, params)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/db/backends/utils.py", line 79, in _execute
    with self.db.wrap_database_errors:
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.IntegrityError: column "tagged_manifest_id" of relation "container_tag" contains null values

(pulp) [vagrant@pulp3-source-fedora35 ~]$ 
(pulp) [vagrant@pulp3-source-fedora35 ~]$ 
(pulp) [vagrant@pulp3-source-fedora35 ~]$ vim devel/pulp_container/pulp_container/app/migrations/0030_enforce_tagged_manifest_reference.py 
(pulp) [vagrant@pulp3-source-fedora35 ~]$ vim devel/pulp_container/pulp_container/app/migrations/0030_enforce_tagged_manifest_reference.py 
(pulp) [vagrant@pulp3-source-fedora35 ~]$ vim devel/pulp_container/pulp_container/app/migrations/0030_enforce_tagged_manifest_reference.py 
(pulp) [vagrant@pulp3-source-fedora35 ~]$ django-admin migrate
Operations to perform:
  Apply all migrations: auth, container, contenttypes, core, file, ostree, sessions
Running migrations:
  Applying container.0030_enforce_tagged_manifest_reference... OK
Access policy for groups/users updated.
Access policy for groups updated.
Access policy for contentguards/core/rbac updated.
Access policy for task-schedules updated.
Access policy for tasks updated.
Access policy for content/container/blobs updated.
Access policy for distributions/container/container updated.
Access policy for pulp_container/namespaces updated.
Access policy for repositories/container/container-push/versions updated.
Access policy for repositories/container/container/versions updated.
Access policy for repositories/container/container-push updated.
Access policy for remotes/container/container updated.
Access policy for repositories/container/container updated.
Access policy for content/container/signatures updated.
Access policy for content/container/manifests updated.
Access policy for content/container/tags updated.
Access policy for acs/file/file updated.
Access policy for content/file/files updated.
Access policy for distributions/file/file updated.
Access policy for publications/file/file updated.
Access policy for remotes/file/file updated.
Access policy for repositories/file/file/versions updated.
Access policy for repositories/file/file updated.
(pulp) [vagrant@pulp3-source-fedora35 ~]$ django-admin shell_plus
Tag.objec# Shell Plus Model Imports
from pulpcore.app.models.access_policy import AccessPolicy, Group
from django.contrib.auth.models import Permission, User
from django.contrib.contenttypes.models import ContentType
from django.contrib.sessions.models import Session
from pulp_container.app.models import Blob, BlobManifest, ContainerDistribution, ContainerNamespace, ContainerPushRepository, ContainerRemote, ContainerRepository, Manifest, ManifestListManifest, ManifestSignature, ManifestSigningService, Tag
from pulpcore.app.models.upload import Upload, UploadChunk
from pulp_file.app.models import FileAlternateContentSource, FileContent, FileDistribution, FilePublication, FileRemote, FileRepository
from pulp_ostree.app.models import OstreeCommit, OstreeCommitObject, OstreeConfig, OstreeDistribution, OstreeObject, OstreeRef, OstreeRemote, OstreeRepository, OstreeSummary
from pulpcore.app.models.acs import AlternateContentSource, AlternateContentSourcePath
from pulpcore.app.models.base import Label
from pulpcore.app.models.content import Artifact, AsciiArmoredDetachedSigningService, Content, ContentArtifact, PulpTemporaryFile, RemoteArtifact, SigningService
from pulpcore.app.models.exporter import Export, ExportedResource, Exporter, FilesystemExport, FilesystemExporter, PulpExport, PulpExporter
from pulpcore.app.models.importer import Import, Importer, PulpImport, PulpImporter, PulpImporterRepository
from pulpcore.app.models.progress import GroupProgressReport, ProgressReport
from pulpcore.app.models.publication import BaseDistribution, ContentGuard, ContentRedirectContentGuard, Distribution, Publication, PublishedArtifact, PublishedMetadata, RBACContentGuard
from pulpcore.app.models.repository import Remote, Repository, RepositoryContent, RepositoryVersion, RepositoryVersionContentDetails
from pulpcore.app.models.role import GroupRole, Role, UserRole
from pulpcore.app.models.status import ContentAppStatus
from pulpcore.app.models.task import CreatedResource, Task, TaskGroup, TaskSchedule, Worker
# Shell Plus Django Imports
from django.core.cache import cache
from django.conf import settings
from django.contrib.auth import get_user_model
from django.db import transaction
from django.db.models import Avg, Case, Count, F, Max, Min, Prefetch, Q, Sum, When
from django.utils import timezone
from django.urls import reverse
from django.db.models import Exists, OuterRef, Subquery
Python 3.10.4 (main, Mar 25 2022, 00:00:00) [GCC 11.2.1 20220127 (Red Hat 11.2.1-9)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.2.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: Tag.objects.all()
Out[1]: <BulkTouchQuerySet [<Tag: latest>]>

```